src/tools/kdb/external.c

```diff
@@ -0,0 +1,205 @@
+/**
+ * @file
+ *
+ * @brief Code to support external programs
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <command.h>
+#include <external.h>
+#include <kdbease.h>
+
+#include <errno.h>
+#include <kdberrors.h>
+#include <spawn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <sys/wait.h>
+#endif
+
+extern char ** environ;
+
+static void insert (StringMap * map, char * key, char * value)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (elektraStrCmp (map->array[i].commandName, key) == 0)
+		{
+			map->array[i].binary = value;
+			return;
+		}
+	}
+	KeyValue newPair;
+	newPair.commandName = key;
+	newPair.binary = value;
+	map->array[map->size++] = newPair;
+}
+
+char * getExternalBin (StringMap * map, const char * key)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (elektraStrCmp (map->array[i].commandName, key) == 0)
+		{
+			return map->array[i].binary;
+		}
+	}
+	return NULL;
+}
+
+void freeExternalBinaryMap (StringMap * map)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (map->array[i].commandName != NULL)
+		{
+			elektraFree (map->array[i].commandName);
+			map->array[i].commandName = NULL;
+		}
+		if (map->array[i].binary != NULL)
+		{
+			elektraFree (map->array[i].binary);
+			map->array[i].binary = NULL;
+		}
+	}
+	map->size = 0;
+}
+
+int tryLoadExternal (char * commandName, StringMap * map)
+{
+	char * execPathPtr = getenv ("KDB_EXEC_PATH");
+	bool found = false;
+	char path[1024] = { 0 };
+	struct stat buf;
+
+	if (execPathPtr)
+	{
+		char * execPath = strdup (execPathPtr);
+		char * token = strtok (execPath, ":");
+		while (token != NULL && !found)
+		{
+			snprintf (path, sizeof (path), "%s/%s", token, commandName);
+			found = stat (path, &buf) != -1;
+			token = strtok (NULL, ":");
+		}
+		elektraFree (execPath);
+	}
+
+	if (!found)
+	{
+		snprintf (path, sizeof (path), "%s/%s", BUILTIN_EXEC_FOLDER, commandName);
+		found = stat (path, &buf) != -1;
+	}
+
+	if (found)
+	{
+		insert (map, elektraStrDup (commandName), strdup (path));
```

Comment:
As an example, with a `KeySet * map`, this would be:

```c
Key * k = keyNew ("/", KEY_VALUE, path, KEY_END);
keyAddBaseName (commandName);
ksAppendKey (map, k);
```